doc/interoperability/pandas.rst

```diff
@@ -0,0 +1,295 @@
+pitfalls and some (not all) of these will be outlined below.
+
+Reasons for these conditions and the difficulty of supporting pandas, stems
+from the fact that pandas data structures such as Series and DataFrames can
+hold heterogenous datatypes (different columns can contain different
+datatypes). Therefore, Series and DataFrames can be thought of as containers
+for arrays that hold the actual data in an homogenous format. Some of these
+underlying data structures might not necessarily be representable as NumPy
+arrays (e.g. Categorical data). Pandas supports these alternate data types using
+pandas specific or 3rd party libraries that extend the NumPy type system. The
+difficulty of supporting pandas data structure thus comes from pandas
+capability of supporting heterogenous data in one data structure.
+
+Every Scikit-learn estimator/transformer/pipeline
+(for the rest of this section we shall call these primitives)
+supports the use of DataFrames as inputs which is achieved by obtaining a
+`NumPy array <https://docs.scipy.org/doc/numpy/user/>`__ using
+the :meth:`~numpy.asarray` on a DataFrame object. The only exception where a
+a DataFrame is being used explicitly is the
+:class:`~sklearn.compose.ColumnTransformer` which is briefly
+discussed below `Dealing with heterogenous data`_.
+
+.. note::
+  Starting with pandas version 0.24.0, it is encouraged to obtain
+  NumPy arrays from DataFrames or Series via :meth:`.to_numpy()` instead of
+  using :meth:`.values`. More details on this can be found in the
+  `release notes <http://pandas-docs.github.io/pandas-docs-travis/whatsnew/v0.24.0.html#accessing-the-values-in-a-series-or-index>`__
+  and the documentation `here <http://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#basics-dtypes>`__
+  and `here <http://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#attributes-and-underlying-data>`__.
+
+There are several conditions on using a DataFrame as an input to
+Scikit-learn primitives, one of which is that the data in the
+DataFrame columns used by the estimator are of numerical type. Other conditions
+and pitfalls are described in subsequent sections. The numerical condition can
+be checked e.g. using something like::
+
+  >>> import pandas as pd
+  >>> from pandas.api.types import is_numeric_dtype
+  >>> df = pd.DataFrame({'foo': [1, 2, 3],
+  ...                    'bar': [1.2, 2.3, 3.]})
+  >>> all([is_numeric_dtype(df[x]) for x in df.columns])
+  True
+  >>> df = pd.DataFrame({'foo': [1,2,3],
+  ...                    'bar': [1.2, 2.3, 3.],
+  ...                    'baz': ['foo', 'bar', 'baz']})
+  >>> all([is_numeric_dtype(df[x]) for x in df.columns])
+  False
+
+There are also a variety of classes such as pipelines and model selection
+processes that will pass a DataFrame along "as is" to the nested estimators
+(see the next section for an example). However, this is not guaranteed and some
```

Comment:
I don't know what "this is not guaranteed" means. For particular estimators, this behaviour is guaranteed, insofar as we guarantee anything.